### PR TITLE
Spring boot security

### PR DESCRIPTION
**Pull request by padu12**

### CODE REVIEW

File: src/main/resources/db/changelog/changes/05-create-users-roles-table.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+                  name: user_id
+                  type: bigint
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
+                    nullable: false
```
Comment (satori11111) [inline src/main/resources/db/changelog/changes/05-create-users-roles-table.yaml:19]: add foreignkeys and i think both user_id and role_id should be primarykey

**Review by satori11111 — APPROVED** [review]
Brilliant! but change your table users_roles

File: pom.xml [hunk 2]
```diff
@@ -27,6 +27,28 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-impl</artifactId>
+            <version>0.11.5</version>
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-jackson</artifactId> <!-- or jjwt-gson if Gson is preferred -->
```
Comment (andrii-hoienko) [inline pom.xml]: Please remove unnecessary comments

File: src/main/java/com/kaziamyr/onlinebookstore/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -2,15 +2,18 @@
...
 import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Size;
 import java.math.BigDecimal;
 import lombok.Data;
 import org.hibernate.validator.constraints.ISBN;
 
 @Data
 public class CreateBookRequestDto {
     @NotNull
+    @Size(min = 1, max = 100)
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/dto/CreateBookRequestDto.java]: Is there are any requirements that title and author should be 100 symbols maximum?

File: src/main/java/com/kaziamyr/onlinebookstore/dto/UserLoginRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+
+@Data
+public class UserLoginRequestDto {
+    @NotNull
+    @Email
+    @Size(min = 1, max = 100)
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/dto/UserLoginRequestDto.java]: Is there are any requirements that password and email should be 100 symbols maximum?

File: src/main/java/com/kaziamyr/onlinebookstore/dto/UserRegistrationRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+public class UserRegistrationRequestDto {
+    @NotNull
+    @Email
+    @Size(min = 1, max = 100)
+    private String email;
+    @NotNull
+    @Size(min = 8, max = 30)
+    private String password;
+    @NotNull
+    private String repeatPassword;
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/dto/UserRegistrationRequestDto.java:21]: Maybe it makes sense to add some constraints here like in password

File: src/main/java/com/kaziamyr/onlinebookstore/model/User.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    private String email;
+    @Column(nullable = false)
+    @ToString.Exclude
+    private String password;
+    @Column(nullable = false)
+    private String firstName;
+    @Column(nullable = false)
+    private String lastName;
+    private String shoppingAddress;
+    @JoinColumn(nullable = false)
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/model/User.java]: Is it necessary to specify JoinColumn along with JoinTable?

File: src/main/java/com/kaziamyr/onlinebookstore/model/User.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    )
+    @ToString.Exclude
+    private Set<Role> roles = new HashSet<>();
+    @Column(nullable = false)
+    private boolean isDeleted;
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        Set<GrantedAuthority> authorities = new HashSet<>();
+
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/model/User.java]: It is not necessary to add line breaks here

File: src/main/java/com/kaziamyr/onlinebookstore/security/CustomUserDetailsService.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+@RequiredArgsConstructor
+@Component
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        return userRepository.findByEmail(username).orElseThrow(
+                () -> new UsernameNotFoundException("Can't find user by email " + username));
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/security/CustomUserDetailsService.java]: cant find by email or by user name?

File: src/main/java/com/kaziamyr/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer")) {
+            return bearerToken.substring(7);
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/security/JwtAuthenticationFilter.java]: magic number

File: src/main/java/com/kaziamyr/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            );
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer")) {
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/security/JwtAuthenticationFilter.java]: magic text

Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/security/JwtAuthenticationFilter.java]: Please ensure that (StringUtils.hasText) makes null check

File: src/main/java/com/kaziamyr/onlinebookstore/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public UserRegistrationResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
+            throw new RegistrationException("User with email "
+                    + request.getEmail() + " is already present!");
+        }
+        User user = new User();
+        user.setPassword(passwordEncoder.encode(request.getPassword()));
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/service/impl/UserServiceImpl.java]: optional: actually it is better to create separate mapper class for such purposes

File: src/main/java/com/kaziamyr/onlinebookstore/validation/FieldMatchValidator.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Objects;
+import org.springframework.util.StringUtils;
+
+public class FieldMatchValidator implements ConstraintValidator<FieldMatch,
+        Object> {
+    private String firstField;
```
Comment (andrii-hoienko) [inline src/main/java/com/kaziamyr/onlinebookstore/validation/FieldMatchValidator.java]: These fields are just first and second? Is there any connection between them?

File: src/main/resources/db/changelog/changes/03-create-roles-table.yaml [hunk 13]
```diff
@@ -0,0 +1,21 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-roles-table
+      author: bohdankaziamyr
```
Comment (andrii-hoienko) [inline src/main/resources/db/changelog/changes/03-create-roles-table.yaml]:
```suggestion
      author: bohdan-kaziamyr
```

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, some minor improvements required

File: src/main/java/com/kaziamyr/onlinebookstore/security/JwtAuthenticationFilter.java (hunk 10 above)
Comment (padu12) [reply src/main/java/com/kaziamyr/onlinebookstore/security/JwtAuthenticationFilter.java]: yes, this method makes null check

image

File: src/main/java/com/kaziamyr/onlinebookstore/mapper/UserMapper.java [hunk 14]
```diff
@@ -3,12 +3,29 @@
...
-    User toModel(UserRegistrationRequestDto requestDto);
+    public abstract UserRegistrationResponseDto toUserResponseDto(User user);
+
+    @BeforeMapping
+    public void addPasswordAndRoles(
+            @MappingTarget User user, UserRegistrationRequestDto requestDto
+    ) {
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        user.setRoles(Set.of(new Role(1L, Role.RoleName.ROLE_USER)));
+    }
```
Comment (nacenik) [inline src/main/java/com/kaziamyr/onlinebookstore/mapper/UserMapper.java]: Bad mapping. It breaks Single responsibility. Put role inside service, not mapper

File: src/main/java/com/kaziamyr/onlinebookstore/model/User.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    private String lastName;
+    private String shoppingAddress;
+    @ManyToMany
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    @ToString.Exclude
+    private Set<Role> roles = new HashSet<>();
```
Comment (nacenik) [inline src/main/java/com/kaziamyr/onlinebookstore/model/User.java:51]: why the field isn't excluded from Equals and hashcode? 
if all fields are the same except roles - are users different or not?

File: pom.xml [hunk 16]
```diff
@@ -27,6 +27,28 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-api</artifactId>
+            <version>0.11.5</version>
+        </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-impl</artifactId>
+            <version>0.11.5</version>
```
Comment (nacenik) [inline pom.xml]: use properties instead of version duplication

**Review by nacenik — CHANGES_REQUESTED** [review]
Should be improved a little bit :)

**Review by andrii-hoienko — APPROVED** [review]
Good job!

**Pull request merged**